validator_client/src/validator_duty.rs

```diff
@@ -53,58 +55,79 @@ impl ValidatorDuty {
         beacon_node: &BeaconNodeHttpClient,
         current_epoch: Epoch,
         request_epoch: Epoch,
-        pubkey: PublicKey,
-    ) -> Result<ValidatorDuty, String> {
-        let pubkey_bytes = PublicKeyBytes::from(&pubkey);
+        mut known_pubkeys: Vec<(PublicKey, u64)>,
+        unknown_pubkeys: &[PublicKey],
+    ) -> Result<Vec<ValidatorDuty>, String> {
+        let mut duties = Vec::new();
 
-        let validator_index = if let Some(index) = beacon_node
-            .get_beacon_states_validator_id(
-                StateId::Head,
-                &ValidatorId::PublicKey(pubkey_bytes.clone()),
-            )
-            .await
-            .map_err(|e| format!("Failed to get validator index: {}", e))?
-            .map(|body| body.data.index)
-        {
-            index
-        } else {
-            return Ok(Self::no_duties(pubkey));
-        };
+        // Query for any pubkeys we don't know the index for in the current epoch.
+        for pubkey in unknown_pubkeys {
+            let pubkey_bytes = PublicKeyBytes::from(pubkey);
+            if let Some(index) = beacon_node
+                .get_beacon_states_validator_id(
+                    StateId::Head,
+                    &ValidatorId::PublicKey(pubkey_bytes.clone()),
+                )
+                .await
+                .map_err(|e| format!("Failed to get validator index: {}", e))?
+                .map(|body| body.data.index)
+            {
+                known_pubkeys.push((pubkey.clone(), index));
+            } else {
+                duties.push(Self::no_duties(pubkey.clone(), None))
+            }
+        }
 
-        if let Some(attester) = beacon_node
-            .get_validator_duties_attester(request_epoch, Some(&[validator_index]))
+        // Query attester duties for known indices, and map the response by index.
+        let attester_data_by_index: HashMap<u64, AttesterData> = beacon_node
+            .post_validator_duties_attester(
+                request_epoch,
+                known_pubkeys.iter().map(|(_, index)| *index).collect(),
+            )
             .await
             .map_err(|e| format!("Failed to get attester duties: {}", e))?
             .data
-            .first()
-        {
-            let block_proposal_slots = if current_epoch == request_epoch {
-                beacon_node
-                    .get_validator_duties_proposer(current_epoch)
-                    .await
-                    .map_err(|e| format!("Failed to get proposer indices: {}", e))?
-                    .data
-                    .into_iter()
-                    .filter(|data| data.pubkey == pubkey_bytes)
-                    .map(|data| data.slot)
-                    .collect()
-            } else {
-                vec![]
-            };
+            .into_iter()
+            .map(|data| (data.validator_index, data))
```

Comment:
I get the feeling that if we mapped into a `ValidatorDuty` here, instead of on L114, we could avoid the `proposal_slots_by_index` map.

It's really not a big deal at all, I just thought I'd mention it.

Reply:
I was able to get rid of the `attester_duties_by_slot` map, but now need to use the pubkey returned by the query for the duty as opposed to the pubkey we have locally.  Not that they should be different, but there's an extra deserialization.  

I also mapped the result of this query to validator indices to compare with indices we queried for so we can create any remaining empty duties.  Let me know if this is along the lines you were thinking.. the codes still a bit ugly, but I'm not really sure how to clean it up further